firmware/components/midi/include/midi.h

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2022 Barabas Raffai
+ *
+ * This file is part of ShrapnelDSP.
+ *
+ * ShrapnelDSP is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * ShrapnelDSP is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * ShrapnelDSP. If not, see <https://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include <assert.h>
+#include <cstdint>
+#include <etl/string_stream.h>
+
+namespace shrapnel {
+namespace midi {
+
+enum MessageType {
+    NOTE_OFF = 0x80,
+    NOTE_ON = 0x90,
+    CONTROL_CHANGE = 0xB0,
+    PROGRAM_CHANGE = 0xC0,
+};
+
+struct Message {
+    typedef uint8_t midi_channel_t;
+
+    struct NoteOnOff {
+        uint8_t note;
+        uint8_t velocity;
+    };
+
+    struct ControlChange {
+        uint8_t control;
+        uint8_t value;
+    };
+
+    struct ProgramChange {
+        uint8_t number;
+    };
+
+    MessageType type;
+    midi_channel_t channel;
+    union {
+        NoteOnOff note_on;
+        NoteOnOff note_off;
+        ControlChange control_change;
+        ProgramChange program_change;
+    };
+
+    friend etl::string_stream& operator<<(etl::string_stream&  out, const Message& message) {
+        out << "{ channel " << message.channel << " ";
+
+        switch(message.type)
+        {
+        case NOTE_ON:
+            out << "note on " << +message.note_on.note << " " << +message.note_on.velocity;
+            break;
+        case NOTE_OFF:
+            out << "note off " << +message.note_on.note << " " << +message.note_on.velocity;
+            break;
+        case CONTROL_CHANGE:
+            out << "control change " << +message.control_change.control << " " << +message.control_change.value;
+            break;
+        case PROGRAM_CHANGE:
+            out << "program change " << +message.program_change.number;
+            break;
+        default:
+            out << +message.type;
+            break;
+        }
+
+        out << "}";
+        return out;
+    }
+};
+
+class Decoder {
+    public:
+    using Callback = std::function<void(Message)>;
```

Comment:
Does `std::function` use the heap?

https://stackoverflow.com/questions/46163607/avoid-memory-allocation-with-stdfunction-and-member-function

Reply:
https://www.etlcpp.com/delegate.html